### ISSUE TITLE
Replace function node in mathematical expression

### ISSUE THREAD

**Issue by spyridon97** [taocpp/PEGTL#267]
I have a mathematical expression such as:

``expression = function123(v1, v2)``

and I want to transform it to:

``expression = (function1(v1, v2) + function2(v1, v2) + function3(v1, v2))``

In order to solve this problem, I tried to utilize regular expressions, such as:

``"function123\\(([^)]+),([^)]+)\\)"``

and replace using the interior tokens to get what I want.

Regular expressions come with certain limitations and problems. that can be identified in the following examples:

```expression = myfunction123(v1, v2) + function123(v1, v2)```

The above example would create problems because ``my`` would be get messed up.

```expression = function123(cross(v1, v2), cross(v2, v1))```

The above example would create problems because the interior nodes would not be able to be extracted.

So basically I understood that regular expressions are not enough to solve this problem so I need something like an ``abstract syntax tree``.

Could you guide me to accomplish what I described?

Thanks a lot in advance.

**Comment by rockola** [comment]
For that particular issue, you can change the regexp so that it will match function123 but not myfunction123: e.g.
"[^A-Za-z0-9]function123" + the rest of the regexp

**Comment by spyridon97** [comment]
> For that particular issue, you can change the regexp so that it will match function123 but not myfunction123: e.g.
> "[^A-Za-z0-9]function123" + the rest of the regexp

That does not work for the second issue though because I want to be able to extract the interior nodes, right?

**Comment by ColinH** [comment]
The first step is to create a proper grammar for the expressions you want to recognise. Regular expressions will not really do the job. Have you looked at `src/example/pegtl/expression.cpp`? The grammar there can parse expressions with function calls:

```
~/Taocpp/pegtl> build/src/example/pegtl/expression "foo(-2+3)*bar(-baz[4]-fraggle(x,y,z)[4])"
Input: foo(-2+3)*bar(-baz[4]-fraggle(x,y,z)[4])
Result: ( foo( ( -( 2 ) + 3 ) ) * bar( ( -( baz[ 4 ] ) - fraggle( x, y, z )[ 4 ] ) ) )
```

In a second step you could for example use `contrib/parse_tree.hpp` to build a syntax tree.

**Comment by ColinH** [comment]
@spyridon97 Hope you managed to see how regular expressions only work for data with rather limited structure, how more general grammars can go beyond, and make some progress. I'll close this issue for now, if you have any related questions feel free to comment anyhow, re-open, or start a new discussion.
